### DISCUSSION TITLE
Formatting "styles" as an alternative to many options

### DISCUSSION THREAD

**Discussion by faultyserver** [biomejs/biome#674] [Ideas]
I'd like to propose an alternative way of configuring the Biome formatter to help reduce bikeshedding over configuration options while also providing enough of an influence to hopefully help more people migrate from Prettier to Biome without changing too much of their existing code.

The new approach is simply two overarching "styles" -- `cozy` and `compact` (names pending) -- which effectively apply a preset of options as they exist in Prettier without actually exposing those options to the user directly. In the future, this approach also lends itself to further changes between the two styles without having to introduce or maintain new options, or worry about all the possible permutations between them.

## Background 

It's a reality that many people and organizations who want to migrate to Biome for formatting are already using Prettier, and likely have been for some time. To that end, a lot of Biome's formatter is based on and formed around compatibility with Prettier's output.

Prettier is a very opinionated formatter, but it does still expose a number of configuration options to influence how code gets formatted. Some of these are "motivatable", meaning they exist for runtime compatibility and integration with other systems, but others are purely cosmetic, like `--bracket-same-line` and `--bracket-spacing`. The fact that these options exist means that projects will inevitably use them, but they also incur a maintenance cost and lead to a slippery slope of adding more over time.

Some time ago, Prettier released an ["Option Philosophy"](https://prettier.io/docs/en/option-philosophy), explaining exactly this conundrum, regretting that some "unmotivatable" options had been added, and stating that new options would not be considered.

Biome has adopted a similar philosophy, and with good reason, to avoid falling down the same path of regret that Prettier is now in. But at the same time, Biome does not currently implement all of those options from Prettier, meaning compatibility will never be perfect. That in itself is okay, but it does complicate adoption of Biome for many projects, who may have a specific option they wish to keep (context: I'm one of those people, and for me that option is `--bracket-same-line` :).

## "Styles" as a better compromise

Styles are overarching "philosophies" about how code should be styled in broad strokes, with the goal of appealing to a person's (or organization's) ideals about what makes "good" and readable code, rather than targeting individual features to be tweaked. A Style rolls many discrete decisions into a single concept that can be more easily understood, explained, and supported as whole. Decisions about nuanced situations can be broken down into "what best fits the philosophy of this Style?", and can help avoid ambiguity or the occasional "magic" behavior (e.g., adding a leading newline in an object can force it to break to multiple lines? sometimes!).

As a comparison to the current configuration options or those of Prettier, a single Style effectively _implies_ the settings of a number of options, without the user having to specify them. Or more specifically, without having to _expose_ the options to the user in the first place. As a result, maintaining a Style will hopefully be less of a burden than having to add and maintain new options over time. They are also more flexible, able to be changed and amended simply by referencing the Style in other places in the formatter, without a need to plumb through additional options or worry about how various options can interplay with each other.

Determining what counts as a Style then becomes the point of debate. I believe that the Option Philosophy adopted from Prettier still stands, and that being an opinionated formatter is still beneficial. Adding too many Styles leads down the same path as adding too many individual options, and keeping the scope limited feels most inline with avoiding that.

I propose starting with two styles that (anecdotally) feel like the most common and compelling philosophies that people have:
- `cozy` - the default. This very nearly matches all of the current defaults and the defaults of Prettier. Bracket spacing is included, JSX closing brackets are put on their own lines, objects are eagerly broken to new lines when a leading newline is present, and so on. The appeal here is that space and alignment provides clarity in the code.
- `compact` - the alternative. This aligns more closely with the _opposite_ of most options in Prettier. Bracket spacing is _omitted_ (see #392 for an example of the difference), JSX closing brackets are kept on the same line as the last attribute, and objects are condensed to a single line unless they cannot fit within the line length limit (or some other heuristic). The appeal here is that density allows for better visual grouping, making distinct "sections" of code more identifiable and readable at a glance.

After taking a quick glance through a GitHub search of prettier configurations, these options seem to often get changed together, leading me to feel like rolling them up into Styles like this can lead to less churn for users looking to migrate from Prettier to a faster tool like Biome.

## Non-stylistic options

Even with Styles, there are still some options which don't fit cleanly on the boundary between `cozy` and `compact`. `trailingComma` is one, where it may need to be set for compatibility. `arrowParens` and `semi` also feel similar, where the goal is consistency regardless of the code structure rather than anything necessarily stylistic (e.g., changing a one-parameter arrow function to a two-parameter function suddenly requires parentheses, or changing a line of code can change automatic semicolon insertion and not require a semicolon where one was needed before, etc.).

For that reason, I believe these options should be maintained separately from Styles, although I recognize that maintaining both may lead to more thrash in the future.

## Language agnosticism

Another great feature of Styles is that they are inherently language-agnostic; `cozy` is not limited to defining formatting features in JavaScript. The chosen Style can also be used when formatting Markdown to determine if prose gets wrapped or not, or additional newlines are added around blocks that may not need them. Or in CSS it could determine when selectors are wrapped to multiple lines, or how values are spread across lines, or where nesting braces are placed.

I think this is a fantastic benefit of Styles as well, as it can entirely eliminate the need for language-specific configuration in many cases, while still also allowing users to use different styles for different segments of their codebases (i.e., `cozy` for JavaScript, but `compact` for Markdown. Or `compact` in this directory but not another, etc.)

## Possible Futures

As mentioned earlier, if Styles are adopted into Biome, they also simplify making stylistic decisions and changes in the future when new cases and nuances arrive. Making a change to styling is no longer limited to "everyone gets this, or else we have to add an option". So long as the decision _feels_ like it maps to the existing Style philosophies and there's some agreement there, it can pretty freely be added.

One example is the idea of breaking or collapsing object literals. Right now this behavior isn't controlled by any option, but the exact written code can influence the behavior if the user adds a leading newline themselves. With Styles, this is something that could be solidified. In `cozy` mode, the existing behavior of intentional wrapping could be kept, but `compact` mode could choose to always collapse or break purely on the line length or another heuristic, avoiding ambiguity and what can appear as "mixed" formatting depending on the user's input.

**Comment by faultyserver** [comment]
## Adding a `minimal` parameter

As mentioned in the "Non-stylistic options" section, there are some configuration options that don't really fit into the philosophy of the two "styles" (cozy and compact): `arrowParens`, `semi`, and `trailingComma`, specifically. Following the goal of reducing how many parameters are configurable, I think a compromise for these options could be a single, separate parameter that controls all of them, which I'll call `minimal`.

The most fundamental decision point in formatting JavaScript realistically comes down to the `semi` option: should all statements be terminated by semicolons, or should we rely on the JS engine to use automatic semi-colon insertion and implicitly understand statement boundaries? I think the underlying philosophy behind that question is whether the formatted code should include "unnecessary" syntax or not. "Semicolons aren't _necessary_ most of the time, so why should they be written in the code?". That question of necessity is why I'm calling this new option `minimal`, to indicate whether _unnecessary syntax_ should be included or removed when formatting.

Taking that philosophy a step further, I think the other options like `arrowParens` and `trailingComma` could reasonably be bundled into the same parameter. Arrow functions with a single named argument don't require parentheses, so why should they be included? And trailing commas don't actually affect the semantics of a list, so why add them?

Of course, all of these questions have a reasonable answer on both sides. A "minimalist" would say they aren't necessary, so they can be left out. But a "maximalist" would likely say they aren't _necessary_, but they add consistency and clarity regardless of the content of the code. A change to the content of an statement makes no difference if we just always add a semicolon after every statement. Always rendering arrow parentheses means changing the parameters of a function won't change the surrounding syntax. And forcing trailing commas reduce line diffs when adding or removing lines to complex expressions.

With all of that, I'll propose having `minimal` be a separate parameter to `style`, _replacing_ all of those remaining configuration options. While `style` affects the general "density" of the formatted code, `minimal` affects the "necessity" of it. In `minimal` mode, any syntax that isn't necessary is removed. Outside of minimal mode, Biome decides whether consistency or conciseness is more important and adds or removes syntax where necessary.

### Note on compatibility

`trailingComma` [exists in prettier](https://prettier.io/docs/en/options.html#trailing-commas) partly because the syntax is sometimes invalid in older browsers and other JS engines, as well as old versions of TypeScript. ES5 doesn't allow trailing commas in argument lists, for example, and will fail. While supporting all engines is nice, I personally think at this point (now 6 years past the general availability of ES2017 which introduced support everywhere) it's safe to trust that it will be valid in most environments. All of the engines and environments that don't support this are now unsupported and end-of-life, and necessary compatibility is still possible (either by ignoring relevant files, downlevel compilation, bundling or other means that are likely already in place for them).

**Comment by faultyserver** [comment]
### `quoteStyle`

Another option that I don't think cleanly fits into _either_ `style` or `minimal` is `quoteStyle` (or the equivalent `singleQuote` option in Prettier).

I'm not entirely sure how to best handle this. By default, Prettier, Biome, and many other tools (e.g., `black` for Python) use a heuristic to determine which quote style is most appropriate (i.e., "limit the number of escaped characters") and only fall back on this option when there is no clear winner there.

I don't think this fits into `style`, because there's no actual change to the density of the code. Arguably a `"` is _more dense_ than a `'`, but in terms of actual screen space they are identical, and the heuristic means it won't be consistent anyway. The same goes for `minimal`, where the necessity of the syntax doesn't change, and no space is saved by using one or the other.

That said, quoting is a strong preference that people have. It doesn't feel _terrible_ to me to remove the option entirely if this approach with `style` is taken, but it would undoubtably cause some turmoil as well.

**Comment by ematipico** [comment]
Even though we introduced most of the prettier lint rules, I believe it's worth pursing this strategy. Even though the Biome formatter is compatible with Prettier, it doesn't mean that we can't provide different **opinionated** visions.

For example, CSS formatting could benefit a `compact` style, where classes and selectors are more cluttered together.

**Reply by Conaclos** [reply]
Regarding JS/TS/JSX formatting, I am a bit skeptical about this proposal. Actually, we do not have that many options. Adding this new way of configuring our formatter could increase the number of options and thus the bug surface. Also, this could potentially confuse users. If we follow this route, we should investigate which options are commonly used together. For example, I think users who set `semi` to `false` don't necessarily set `arrowParens` to `asNeeded` and `trailingCommas` to `none`.

We could re-evaluate this strategy once the CSS formatting is in place.

**Reply by faultyserver** [reply]
I definitely want to take a look at what the most common configurations are. Like somehow polling the top 5k repositories with `.prettierrc`/`prettier.config.js` and such and seeing what they use. That would probably help shape a lot of the thinking about what's actually a useful option.

**Comment by GermanJablo** [comment]
> ...objects are condensed to a single line unless they cannot fit within the line length limit (or some other heuristic). The appeal here is that density allows for better visual grouping, making distinct "sections" of code more identifiable and readable at a glance. @faultyserver 

> ...CSS formatting could benefit a compact style, where classes and selectors are more cluttered together. @ematipico 

I couldn't agree more with both statements! I am definitely `compact` team. I might be interested in trying `minimal`, but having at least `compact` would be a killer feature for me.

I love that opinionated formatters like Prettier and Biome avoid us from discussions of little relevance. Our brains get used to different styles, like using `'` vs `"`, or using trailing commas. Whatever you choose, you will get used to it in a matter of time.

However, I am convinced that rules that use extra lines unnecessarily are **objectively** less readable.

I just don't think anyone in their right mind would prefer this:

```css
h1 {
  @apply text-default text-2xl font-bold;
}
h2 {
  @apply text-default text-xl font-bold;
}
h3 {
  @apply text-default text-lg font-bold;
}
h4 {
  @apply text-default text-base font-bold;
}
h5 {
  @apply text-default text-sm font-bold;
}
h6 {
  @apply text-default text-xs font-bold;
}
```
to this: 

```css
h1 { @apply text-2xl font-bold text-default; }
h2 { @apply text-xl font-bold text-default; }
h3 { @apply text-lg font-bold text-default; }
h4 { @apply text-base font-bold text-default; }
h5 { @apply text-sm font-bold text-default; }
h6 { @apply text-xs font-bold text-default; }
```

Or that he/she may prefer this:

```js
if (foo === 1) {
  return "one";
}
if (foo === 2) {
  return "two";
}
if (foo === 3) {
  return "three";
}
if (foo === 4) {
  return "four";
}
if (foo === 5) {
  return "five";
}
```

to this:

```js
if (foo === 1) return "one";
if (foo === 2) return "two";
if (foo === 3) return "three";
if (foo === 4) return "four";
if (foo === 5) return "five";
```

This is why I have to use `// prettier-ignore` all the time in my code 🫠
